### PR TITLE
Swift: Add predicate injection query

### PR DESCRIPTION
**Pull request by atorralba**

### CODE REVIEW

File: swift/ql/src/queries/Security/CWE-943/PredicateInjection.ql [hunk 1]
```diff
@@ -0,0 +1,22 @@
+/**
+ * @name Predicate built from user-controlled sources
+ * @description Building a NSPredicate from user-controlled sources may lead to attackers
```
Comment (geoffw0) [inline swift/ql/src/queries/Security/CWE-943/PredicateInjection.ql:3]: Should this be
```suggestion
 * @description Building an NSPredicate from user-controlled sources may lead to attackers
```
?

File: swift/ql/test/query-tests/Security/CWE-946/predicateInjection.swift [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+func test() {
+    let remoteString = String(contentsOf: URL(string: "http://example.com/")!)
+
+    NSPredicate(format: remoteString, argumentArray: []) // $ hasPredicateInjection=23
+    NSPredicate(format: remoteString, arguments: CVaListPointer(_fromUnsafeMutablePointer: UnsafeMutablePointer(bitPattern: 0)!)) // $ hasPredicateInjection=23
+    NSPredicate(format: remoteString) // $ hasPredicateInjection=23
+    NSPredicate(format: remoteString, "" as! CVarArg) // $ hasPredicateInjection=23
+    NSPredicate(fromMetadataQueryString: remoteString) // $ hasPredicateInjection=23
+}
```
[the remark this answers is not in the document]
Comment (atorralba) [reply swift/ql/test/query-tests/Security/CWE-946/predicateInjection.swift:36]: Good point! Added in https://github.com/github/codeql/pull/11670/commits/8cda23a2b8433dae7e7ec9598a058005cac88da5.

**Review by geoffw0 — APPROVED** [review]
LGTM.  Happy for this to be merged after docs review.

File: swift/ql/src/queries/Security/CWE-943/PredicateInjection.qhelp [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+Predicates represent logical conditions that can be used to check whether an object matches them.
+If a predicate is built from user-provided data without sufficient sanitization, an attacker may
+be able to change the overall meaning of the predicate.
+</p>
+</overview>
+<recommendation>
+<p>
+When building a predicate from untrusted data, pass it to the appropriate <code>arguments</code> parameter
+during initialization, or as an array of substitution variables before evaluation, instead of appending
+or concatenating it to the body of the predicate.
```
Comment (subatoi) [inline swift/ql/src/queries/Security/CWE-943/PredicateInjection.qhelp:14]:
```suggestion
When building a predicate from untrusted data, you should either pass it to the appropriate <code>arguments</code> parameter during initialization, or as an array of substitution variables before evaluation. You should not append or concatenate it to the body of the predicate.
```

Just a suggestion to break up the sentence slightly

File: swift/ql/src/queries/Security/CWE-943/PredicateInjection.qhelp [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+</p>
+</recommendation>
+<example>
+<p>
+In the following insecure example, a <code>NSPredicate</code> is built directly from data obtained from an HTTP request,
+which is untrusted, and can be arbitrarily set by an attacker to alter the meaning of the predicate:
```
Comment (subatoi) [inline swift/ql/src/queries/Security/CWE-943/PredicateInjection.qhelp:20]:
```suggestion
In the following insecure example, <code>NSPredicate</code> is built directly from data obtained from an HTTP request. This is untrusted, and can be arbitrarily set by an attacker to alter the meaning of the predicate:
```

Again, just a minor suggestion to break up the sentence slightly.

File: swift/ql/src/queries/Security/CWE-943/PredicateInjection.qhelp [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+</p>
+<sample src="PredicateInjectionBad.swift" />
+<p>
+A better way to do this is using the <code>arguments</code> parameter of <code>NSPredicate</code>'s constructor,
+which prevents attackers from altering the meaning of the predicate, even if they control the externally-obtained data,
+as seen in the following secure example:
```
Comment (subatoi) [inline swift/ql/src/queries/Security/CWE-943/PredicateInjection.qhelp:26]:
```suggestion
A better way to do this is to use the <code>arguments</code> parameter of <code>NSPredicate</code>'s constructor. This prevents attackers from altering the meaning of the predicate, even if they control the externally obtained data, as seen in the following secure example:
```

Minor suggestions: mainly sentence break-up again but also a very nitpicky suggestion as compound adjectives using `-ly` adverbs don't conventionally use hyphens.

File: swift/ql/src/queries/Security/CWE-943/PredicateInjection.qhelp [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+</p>
+<sample src="PredicateInjectionGood.swift" />
+</example>
+<references>
+<li>Apple Developer documentation: <a href="https://developer.apple.com/documentation/foundation/nspredicate">NSPredicate</a> </li>
```
Comment (subatoi) [inline swift/ql/src/queries/Security/CWE-943/PredicateInjection.qhelp:31]:
```suggestion
<li>Apple Developer Documentation: <a href="https://developer.apple.com/documentation/foundation/nspredicate">NSPredicate</a> </li>
```

Minor: I guess if we're capitalizing `Developer` we should capitalize `Documentation`, too, as Apple themselves seem to.

**Review by subatoi — COMMENTED** [review]
👋 from Docs—just a few minor suggestions!

**Comment by atorralba** [comment]
Thanks for the review @subatoi! 🙇 All suggestions applied.

**Review by subatoi — APPROVED** [review]
Thanks @atorralba ! In the unlikely event you need anyone else from Docs to review this again after tomorrow and before the 3rd January, please post in #code-security-docs and someone should be able to help. Have a good break!

**Review by geoffw0 — APPROVED** [review]

**Pull request merged**